ogusa/parameter_plots.py

```diff
@@ -834,3 +866,133 @@ def plot_income_data(ages, abil_midp, abil_pcts, emat, output_dir=None,
+    assert (start_year <= TC_LAST_YEAR)
+    assert (year <= TC_LAST_YEAR)
+    assert (year >= start_year)
+    # if list of tax function types less than list of params, assume
+    # all the same functional form
+    if len(tax_func_type) < len(tax_param_list):
+        tax_func_type = [tax_func_type[0]] * len(tax_param_list)
+    for i, v in enumerate(tax_func_type):
+        assert (v in ['DEP', 'DEP_totalinc', 'GS', 'linear'])
+    assert (rate_type in ['etr', 'mtrx', 'mtry'])
+    assert (len(tax_param_list) == len(labels))
+
+    # Set age and year to look at
+    if age is not None:
+        assert isinstance(age, int)
+        s = age - 21
+    else:
+        s = 0  # if not age-specific, all ages have the same values
+    t = year - start_year
+
+    # create rate_key to correspond to keys in tax func dicts
+    rate_key = 'tfunc_' + rate_type + '_params_S'
+
+    # Set income range to plot over (min income value hard coded to 5)
+    inc_sup = np.exp(np.linspace(np.log(5), np.log(max_inc_amt), 100))
+    # Set income value for other income
+    inc_fix = other_inc_val
+
+    if over_labinc:
+        key1 = 'total_labinc'
+        X = inc_sup
+        Y = inc_fix
+    else:
+        key1 = 'total_capinc'
+        X = inc_fix
+        Y = inc_sup
+
+    # get tax rates for each point in the income support and plot
+    fig, ax = plt.subplots()
+    for i, tax_params in enumerate(tax_param_list):
+        rates = txfunc.get_tax_rates(
+            tax_params[rate_key][s, t, :], X, Y, None, tax_func_type[i],
+            rate_type, for_estimation=False)
+        plt.plot(inc_sup, rates, label=labels[i])
+
+    # plot raw data (if passed)
+    if data_list is not None:
+        # rate_type_dict = {'etr': 'etr', 'mtrx': 'mtr_labinc',
+        #                   'mtry': 'mtr_capinc'}
+        # censor data to range of the plot
+        for d, data in enumerate(data_list):
+            data_to_plot = data[str(year)].copy()
+            if age is not None:
+                data_to_plot.drop(
+                    data_to_plot[data_to_plot['age'] != age].index,
+                    inplace=True)
+            # other censoring 
+            data_to_plot.drop(
+                data_to_plot[data_to_plot[key1] > max_inc_amt].index,
+                inplace=True)
+            # other censoring used in txfunc.py
+            data_to_plot = txfunc.tax_data_sample(data_to_plot)
+            # set number of bins to 100 or bins of $1000 dollars
+            n_bins = min(100, np.floor_divide(max_inc_amt, 1000))
+            # need to compute weighted averages by group...
+            wm = lambda x: np.average(
+                x, weights=data_to_plot.loc[x.index, "weight"])
+            data_to_plot['inc_bin'] = pd.cut(data_to_plot[key1], n_bins)
+            groups = pd.DataFrame(data_to_plot.groupby(["inc_bin"]).agg(
+                rate=(rate_type, wm), income=(key1, wm)))
+            plt.scatter(groups['income'], groups['rate'], alpha=0.1)
```

Comment:
WDYT of adjusting alpha based on the number of tax units in the group? Right now it looks like 1 gets the same visual representation as 1M (is that right?) which makes it harder to assess the fit.

Reply:
@MaxGhenis One reason I want to keep alpha low is because on can pass a list of data objects (e.g., one for each of several specifications of policy) and I want one to be able to see those points and the tax functions easily.  Also, I think it could be distracting to have some dots for the same data series that are lighter or darker in color.   

I view these graphs as a way to quickly diagnose issues with shape and fit (e.g., see Issue #635).  If one wants more detail to see what's going on, it's probably best to look at statistical measures of fit, such as the SSE's that are stored in the tax function estimation output.